<a href="https://colab.research.google.com/github/MariaJoseBoteroZapata/Business-Intelligence/blob/main/Parcial_1_Maria_Jose_Botero.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

0. Cargar las librerias de trabajo

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.cluster import KMeans
from sklearn.metrics import confusion_matrix

from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


#**Descripción de las variables**#
La base de datos del archivo NBA, posee un total de 445 datos. Dentro de estos se encuentran 9 variables:

•	Salary: Es una variable socioeconómica que representa el salario anual de cada jugador. También puede indicar la experiencia que tenga y sus habilidades.

•	Age: Es la edad de los jugadores, esta puede reflejar el estado de un jugador, pues si es mayor logra más experiencia en el baloncesto; sin embargo, si es menor podría tener más agilidad y potencial.

•	Height: Hace referencia a la altura de los jugadores. Esta nos puede decir si un jugador tiene más ventajas para el equipo, ya que, a mayor altura, tendrá más habilidad y capacidad de defensa.

•	Weight: Es el peso de cada jugador, puede repercutir en el desempeño del juego, ya que cambia la condición física.

•	games_started: Son los juegos iniciados, el cual indica el número de juegos en donde un jugador fue titular, además de mostrar la trayectoria e importancia que tiene en el equipo.

•	minutes_per_game: Son los minutos por juego, lo cual indica el promedio en minutos que un jugador está en el partido.

•	Assits: Número de asistencias por jugador, dice la capacidad del jugador para crear nuevas anotaciones con su equipo.

•	Blocks: Número de bloqueos que hace el jugador en el partido. Indica la capacidad de defensa hacia el equipo contrario.

•	Steals: Número de robos del balón que hace cada jugador. Muestra la capacidad para recuperar el balón del equipo oponente.

•	games_played: Es el total de partidos jugados que un jugador ha realizado en la temporada.





1. Se carga el archivo y se seleccionan las variables

In [2]:
nxl='/content/drive/MyDrive/Colab Notebooks/4. BD2_NBA_Data_M_Int.xlsx'
XDB=pd.read_excel(nxl)
XDB=XDB[['salary','age','height','weight','games_started','minutes_per_game','assists','blocks','steals','games_played']]
XDB=XDB.dropna()
display(XDB)


salary  age  height  weight  games_started  minutes_per_game  assists  \
0      947276   36      79     260            838              32.4      2.7   
1    25000000   37      78     212           1198              36.1      4.7   
2     4088019   39      78     220            954              32.4      3.4   
3     5675000   36      77     195            432              30.9      3.6   
4     5250000   40      83     250           1389              34.0      3.0   
..        ...  ...     ...     ...            ...               ...      ...   
440   1733040   19      78     202              6              14.3      0.6   
441   1140240   21      83     200              4              15.1      0.4   
442   1131960   20      81     220              0               4.2      0.0   
443   3102240   20      77     200             66              30.4      5.5   
444    525093   23      79     185              0               3.5      0.1   

     blocks  steals  games_played  
0       0.5     1.8           966  
1       0.5     1.4          1346  
2       0.6     1.1          1274  
3       0.2     1.0          1100  
4       2.2     0.7          1392  
..      ...     ...           ...  
440     0.2     0.4            70  
441     0.5     1.2            24  
442     0.0     0.0             5  
443     0.5     1.0            68  
444     0.0     0.0            12  

[445 rows x 10 columns]

#**Implementación del modelo Naive Bayes**

1. Clasificar los jugadores con el promedio de la variable "Games_played"

In [3]:
import pandas as pd

# Calcular el promedio de la columna 'games_played'
games_played_mean = XDB['games_played'].mean()

# Crear una nueva columna 'above_below_avg'
XDB['above_below_avg'] = XDB['games_played'].apply(lambda x: 1 if x > games_played_mean else 0)

# Mostrar el DataFrame con la nueva columna
print(XDB)

       salary  age  height  weight  games_started  minutes_per_game  assists  \
0      947276   36      79     260            838              32.4      2.7   
1    25000000   37      78     212           1198              36.1      4.7   
2     4088019   39      78     220            954              32.4      3.4   
3     5675000   36      77     195            432              30.9      3.6   
4     5250000   40      83     250           1389              34.0      3.0   
..        ...  ...     ...     ...            ...               ...      ...   
440   1733040   19      78     202              6              14.3      0.6   
441   1140240   21      83     200              4              15.1      0.4   
442   1131960   20      81     220              0               4.2      0.0   
443   3102240   20      77     200             66              30.4      5.5   
444    525093   23      79     185              0               3.5      0.1   

     blocks  steals  games_played  abov

2. Se seleccionan las variables de entrada y salida

In [4]:
#Se separan las variables de entrada y salida
XD=XDB[['salary','age','height','weight','games_started','minutes_per_game','assists','blocks','steals']] #Variables de Entrada
yd=XDB[['above_below_avg']] #Variable de Salida - Variable de Pronóstico

3. Número de datos

In [5]:
num_datos = len(XD)
print(num_datos)

445


4. Porcentaje de datos por clasificación

In [6]:
category_counts = XDB['above_below_avg'].value_counts()
category_percentages = (category_counts / len(XDB)) * 100
print(category_percentages)

above_below_avg
0    57.303371
1    42.696629
Name: count, dtype: float64


5. Se llama al modelo

In [7]:
from sklearn.naive_bayes import GaussianNB

mnb=GaussianNB ()
mnb.fit(XD,yd)

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


GaussianNB()

6. Media, Varianza, Desviación y límites de las ocho variables


In [8]:
#Obtenemos la información del modelo
u=mnb.theta_ #Obtenemos la media
print("la media del Salario,Edad,Altura,Peso,Juegos iniciados,Minutos por juego,Asistencia, Bloques, Robos : \n", u)

var=mnb.var_
print("La varianza del Salario,Edad,Altura,Peso,Juegos iniciados,Minutos por juego,Asistencia, Bloques, Robos : \n", var)
sigma=np.sqrt(var)
print("La desviación estándar de la Salario,Edad,Altura,Peso,Juegos iniciados,Minutos por juego,Asistencia, Bloques, Robos : \n", sigma)

#Se determinan los límites superiores e inferiores de las tres variables
LS=u+sigma
LI=u-sigma

print("El limite superior de las ocho variables es \n", LS)
print("El limite inferior de las ocho variables es \n", LI)

la media del Salario,Edad,Altura,Peso,Juegos iniciados,Minutos por juego,Asistencia, Bloques, Robos : 
 [[2.91582743e+06 2.41607843e+01 7.91647059e+01 2.19423529e+02
  6.52352941e+01 1.82878431e+01 1.46941176e+00 3.88627451e-01
  5.98431373e-01]
 [7.42986398e+06 3.04684211e+01 7.92684211e+01 2.24731579e+02
  4.09294737e+02 2.66994737e+01 2.51473684e+00 5.62105263e-01
  8.41578947e-01]]
La varianza del Salario,Edad,Altura,Peso,Juegos iniciados,Minutos por juego,Asistencia, Bloques, Robos : 
 [[1.15798849e+13 2.73523678e+04 2.73566841e+04 2.80171672e+04
  3.39158480e+04 2.74045416e+04 2.73465659e+04 2.73449991e+04
  2.73449589e+04]
 [3.68265886e+13 2.73570701e+04 2.73564280e+04 2.80246595e+04
  1.10096418e+05 2.73810055e+04 2.73484892e+04 2.73450417e+04
  2.73449657e+04]]
La desviación estándar de la Salario,Edad,Altura,Peso,Juegos iniciados,Minutos por juego,Asistencia, Bloques, Robos : 
 [[3.40292300e+06 1.65385513e+02 1.65398561e+02 1.67383294e+02
  1.84162559e+02 1.65543171e+02 1.653

7. Se evalua el desempeño del modelo

In [9]:
from sklearn.metrics import confusion_matrix

ydp2=mnb.predict(XD)
cm=confusion_matrix(yd,ydp2)
print("La matriz de confusión es:\n",cm)

VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]
Exactitud=(VP+VN)/(VP+VN+FP+FN)
Sensibilidad=VP/(VP+FN)
Especificidad=VN/(VN+FP)
Precisión=VP/(VP+FP)
TasaError=(FP+FN)/(VP+VN+FP+FN)
print("La Exactitud es:",Exactitud)
print("La Sensibilidad es:",Sensibilidad)
print("La Especificidad es:",Especificidad)
print("La Precisión es:",Precisión)
print("La Tasa de Error es:",TasaError)
PredNeg=VN/(VN+FP)


La matriz de confusión es:
 [[236  19]
 [ 76 114]]
La Exactitud es: 0.7865168539325843
La Sensibilidad es: 0.6
La Especificidad es: 0.9254901960784314
La Precisión es: 0.8571428571428571
La Tasa de Error es: 0.21348314606741572


8. Determinar la categoría a la que pertenece una persona con las siguientes características mediante la utilización del modelo Naive Bayes:

In [10]:
XDS=[[5000000,37,82,245,45,17.8,0.5,1.4,0.4]]
ydp= mnb.predict(XDS) #Se obtiene la predicción
print("La predicción del jugador es:\n",ydp)

if ydp==0:
  print("El jugador de la NBA está por debajo del promedio")

if ydp==1:
  print("El jugador de la NBA está por encima del promedio")

La predicción del jugador es:
 [0]
El jugador de la NBA está por debajo del promedio


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


#**Árboles de decisión**

1. Seleccionamos las variables de entrada y salida

In [11]:
nxl='/content/drive/MyDrive/Colab Notebooks/4. BD2_NBA_Data_M_Int.xlsx'
XDB=pd.read_excel(nxl)
XDB=XDB[['salary','age','height','weight','games_started','minutes_per_game','assists','blocks','steals','games_played']]
XDB=XDB.dropna()
display(XDB)

# Calcular el promedio de la columna 'games_played'
games_played_mean = XDB['games_played'].mean()

# Crear una nueva columna 'above_below_avg'
XDB['above_below_avg'] = XDB['games_played'].apply(lambda x: 1 if x > games_played_mean else 0)

#Se separan las variables de entrada y salida
XD=XDB[['salary','age','height','weight','games_started','minutes_per_game','assists','blocks','steals']] #Variables de Entrada
yd=XDB[['above_below_avg']] #Variable de Salida - Variable de Pronóstico

salary  age  height  weight  games_started  minutes_per_game  assists  \
0      947276   36      79     260            838              32.4      2.7   
1    25000000   37      78     212           1198              36.1      4.7   
2     4088019   39      78     220            954              32.4      3.4   
3     5675000   36      77     195            432              30.9      3.6   
4     5250000   40      83     250           1389              34.0      3.0   
..        ...  ...     ...     ...            ...               ...      ...   
440   1733040   19      78     202              6              14.3      0.6   
441   1140240   21      83     200              4              15.1      0.4   
442   1131960   20      81     220              0               4.2      0.0   
443   3102240   20      77     200             66              30.4      5.5   
444    525093   23      79     185              0               3.5      0.1   

     blocks  steals  games_played  
0       0.5     1.8           966  
1       0.5     1.4          1346  
2       0.6     1.1          1274  
3       0.2     1.0          1100  
4       2.2     0.7          1392  
..      ...     ...           ...  
440     0.2     0.4            70  
441     0.5     1.2            24  
442     0.0     0.0             5  
443     0.5     1.0            68  
444     0.0     0.0            12  

[445 rows x 10 columns]

2. Se procede con la implementación del modelo

In [12]:
#Librerias Árbol
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
mar.fit(XD,yd)

#Se procede con la gráfica
from pydotplus import graph_from_dot_data #Me crea la gráfica
from sklearn.tree import export_graphviz #Me exporta a PNG,JPEG,TIFF

ve=['salary','age','height','weight','games_started','minutes_per_game','assists','blocks','steals'] #Nombres para el Árbol
dot_data=export_graphviz(mar,feature_names=ve) #Crear el gráfico
graph=graph_from_dot_data(dot_data)
graph.write_png('2. ÁrbolDecisión.png') #Guardar la gráfica

True

3. Número de nodos puros

In [13]:
import numpy as np
from sklearn.tree import DecisionTreeClassifier

# Asumiendo que 'mar' es tu modelo de árbol de decisión entrenado

# Obtener el número de nodos en el árbol
n_nodes = mar.tree_.node_count

# Obtener los valores de los nodos (número de muestras por clase)
values = mar.tree_.value

# Contar los nodos puros
pure_node_count = 0
for i in range(n_nodes):
    # Verificar si el nodo es una hoja (no tiene hijos)
    if mar.tree_.children_left[i] == mar.tree_.children_right[i]:
        # Verificar si el nodo es puro (uno de los valores es 0)
        if values[i][0][0] == 0 or values[i][0][1] == 0:
            pure_node_count += 1

print("Número de nodos puros:", pure_node_count)

Número de nodos puros: 6


3. Se procede con el pronóstico de los jugadores de la NBA que no cumplen con el promedio

In [14]:
# Cargar el archivo Excel con dos hojas
Pronosticar_datos = XDB

XD = XDB[['salary','age','height','weight','games_started','minutes_per_game','assists','blocks','steals']]
yd = XDB[['above_below_avg']]


X_pronostico = Pronosticar_datos[['salary','age','height','weight','games_started','minutes_per_game','assists','blocks','steals']]

predicciones = mar.predict(X_pronostico)
Pronosticar_datos['Predictions'] = predicciones

# Guardar los resultados en un archivo Excel con una nueva columna de predicciones
Pronosticar_datos.to_excel('pronosticojugadores.xlsx', index=False)

print("Pronóstico completado y guardado en 'pronosticojugadores.xlsx'")

Pronóstico completado y guardado en 'pronosticojugadores.xlsx'


**4**. Evaluación del modelo

In [15]:
ydp=mar.predict(XD) #Pronóstico sobre la base de datos original
cm=confusion_matrix(yd,ydp) #Matriz de confusión
print("La matriz de confusión es:\n",cm)

VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]


#Métricas de desempeño
Ex=(VP+VN)/(VP+VN+FP+FN)
print("La exactitud del modelo es:\n",Ex)
TE = (FP + FN) / (VP + VN + FP + FN)
print("La tasa de error del modelo es:\n",TE)

Se= VP/(VP+FN)
print("La sensibilidad del modelo es:\n",Se)

Sp=VN/(VN+FP)
print("La especificidad del modelo es:\n",Sp)

Pr= VP/(VP+FP)
print("La precisión del modelo es:\n",Pr)

PredNeg= VN/(VN+FN)
print("La precisión negativa del modelo es:\n",PredNeg)

La matriz de confusión es:
 [[247   8]
 [ 12 178]]
La exactitud del modelo es:
 0.9550561797752809
La tasa de error del modelo es:
 0.0449438202247191
La sensibilidad del modelo es:
 0.9368421052631579
La especificidad del modelo es:
 0.9686274509803922
La precisión del modelo es:
 0.956989247311828
La precisión negativa del modelo es:
 0.9536679536679536


5. De acuerdo con la estructura del árbol de decisión, determinar la regla que cumple el deportista y cual es la probabilidad de pertenencia.

In [16]:
import pandas as pd
from sklearn.tree import _tree

# Datos del jugador en el mismo orden que las características usadas en el modelo
J1 = np.array([[5000000, 37, 82, 245, 45, 17.8, 0.5, 1.4, 0.4]])

# Predicción de la clase del jugador
prediccion = mar.predict(J1)

# Probabilidad de pertenencia a cada clase
probabilidades = mar.predict_proba(J1)

# Mostrar resultados
print(f"Predicción del jugador: {prediccion[0]}")
print(f"Probabilidad de estar por debajo del promedio: {probabilidades[0][0]:.2%}")
print(f"Probabilidad de estar por encima del promedio: {probabilidades[0][1]:.2%}")

# Obtener el camino recorrido en el árbol por el jugador
nodos_recorridos = mar.decision_path(J1)

# Obtener las reglas del árbol
nodos = mar.tree_
caracteristicas = mar.tree_.feature
umbrales = mar.tree_.threshold

print("Reglas seguidas por el jugador:")

# Recorrer los nodos activados por el jugador
for indice_nodo in nodos_recorridos.indices:
    if indice_nodo < len(caracteristicas) and caracteristicas[indice_nodo] != -2:
        print(f"Si {ve[caracteristicas[indice_nodo]]} <= {umbrales[indice_nodo]:.2f}")

Predicción del jugador: 1
Probabilidad de estar por debajo del promedio: 16.67%
Probabilidad de estar por encima del promedio: 83.33%
Reglas seguidas por el jugador:
Si age <= 26.50
Si games_started <= 84.00
Si games_started <= 38.00
Si assists <= 0.65


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but DecisionTreeClassifier was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but DecisionTreeClassifier was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but DecisionTreeClassifier was fitted with feature names
  warnings.warn(


#**Modelo K-Means(Clusterización)**

1. Seleccionamos las variables de entrada y salida

In [17]:
nxl='/content/drive/MyDrive/Colab Notebooks/4. BD2_NBA_Data_M_Int.xlsx'
XDB=pd.read_excel(nxl)
XDB=XDB[['salary','age','height','weight','games_started','minutes_per_game','assists','blocks','steals','games_played']]
XDB=XDB.dropna()
display(XDB)

# Calcular el promedio de la columna 'games_played'
games_played_mean = XDB['games_played'].mean()

# Crear una nueva columna 'above_below_avg'
XDB['games_played'] = XDB['games_played'].apply(lambda x: 1 if x > games_played_mean else 0)

#Se separan las variables de entrada y salida
XD=XDB[['salary','age','height','weight','games_started','minutes_per_game','assists','blocks','steals']] #Variables de Entrada
yd=XDB[['games_played']] #Variable de Salida - Variable de Pronóstico
#Se separan las variables de entrada y salida
XD=XDB[['salary','age','height','weight','games_started','minutes_per_game','assists','blocks','steals']] #Variables de Entrada
yd=XDB[['games_played']] #Variable de Salida - Variable de Pronóstico

salary  age  height  weight  games_started  minutes_per_game  assists  \
0      947276   36      79     260            838              32.4      2.7   
1    25000000   37      78     212           1198              36.1      4.7   
2     4088019   39      78     220            954              32.4      3.4   
3     5675000   36      77     195            432              30.9      3.6   
4     5250000   40      83     250           1389              34.0      3.0   
..        ...  ...     ...     ...            ...               ...      ...   
440   1733040   19      78     202              6              14.3      0.6   
441   1140240   21      83     200              4              15.1      0.4   
442   1131960   20      81     220              0               4.2      0.0   
443   3102240   20      77     200             66              30.4      5.5   
444    525093   23      79     185              0               3.5      0.1   

     blocks  steals  games_played  
0       0.5     1.8           966  
1       0.5     1.4          1346  
2       0.6     1.1          1274  
3       0.2     1.0          1100  
4       2.2     0.7          1392  
..      ...     ...           ...  
440     0.2     0.4            70  
441     0.5     1.2            24  
442     0.0     0.0             5  
443     0.5     1.0            68  
444     0.0     0.0            12  

[445 rows x 10 columns]

2. Implementación del modelo

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


#Se procede con la implementación del modelo

from sklearn.cluster import KMeans
from sklearn.metrics import confusion_matrix


#Para mirar correlaciones y variables importantes
sns.pairplot(XDB,hue='games_played',size=4,vars=['salary','age','height','weight','games_started','minutes_per_game','assists','blocks','steals'])
plt.show()

#¿Y cuáles son las correlaciones?
plt.figure()
sns.heatmap(XDB.corr(),annot=True)
plt.show()

/usr/local/lib/python3.11/dist-packages/seaborn/axisgrid.py:2100: UserWarning: The `size` parameter has been renamed to `height`; please update your code.
  warnings.warn(msg, UserWarning)


3. Se procede con la implementación del modelo de clusterización.

In [ ]:
XD=XDB[['salary','age','height','weight','games_started','minutes_per_game','assists','blocks','steals']] #Variables de Entrada
yd=XDB[['games_played']]


#Se van a crear cinco montoncitos de datos (perfiles de crédito)
np.random.seed(42)
nc=5
mkm=KMeans(n_clusters=5,init="random", random_state=42)
mkm.fit(XD,yd)
centroids=mkm.cluster_centers_
labels=mkm.labels_
print("Los centroides son:\n",centroids)

#Como quedaron clasificadas las personas
labels=mkm.labels_
XDB['labels']=labels #Adicione a la tabla el cluster al que pertenece un solicitante
display(XDB)

#Caracterizamos los centroides
ND=np.bincount(labels)
print("El tamaño de los clusters es:\n",ND)
PJPP=XDB.groupby('labels').agg({"games_played":lambda x:(x==1).sum()})
PJPDP=XDB.groupby('labels').agg({"games_played":lambda x:(x==0).sum()})
print("El numero de aprobados por cluster es:\n",PJPP)
print("El numero de negados por cluster es:\n",PJPDP)

#Se procede con la estimación de los porcentajes de preaprobación
PJPP['games_played']=PJPP=PJPP['games_played']/ND
PJPDP['games_played']=PJPDP=PJPDP['games_played']/ND
print("El porcentaje de aprobados por cluster es:\n",PJPP)
print("El porcentaje de negados por cluster es:\n",PJPDP)

4. Evaluación del modelo

In [ ]:
ydp=np.zeros((len(XD),1))
PJPP=np.array(PJPP);PJPDP=np.array(PJPDP)

for k in range(len(XD)):
  pjpp=labels[k]
  print("La persona tal:",k)
  print("El cluster al que pertenece el jugador es:",PJPP)
  print("El procentaje de Jugadores por encima del promedio es:",PJPP[pjpp])
  print("El procentaje de Jugadores por debajo del promedio es:",PJPDP[pjpp])

  if PJPP[pjpp]>PJPDP[pjpp]:
    ydp[k]=1
  else:
    ydp[k]=0



5. Se procede a evaluar el modelo (matriz de confusión)

In [ ]:
cm=confusion_matrix(yd,ydp)
print("La matriz de confusión es:\n",cm)

VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]
Exactitud=(VP+VN)/(VP+VN+FP+FN)
Sensibilidad=VP/(VP+FN)
Especificidad=VN/(VN+FP)
Precisión=VP/(VP+FP)
TasaError=(FP+FN)/(VP+VN+FP+FN)
print("La Exactitud es:",Exactitud)
print("La Sensibilidad es:",Sensibilidad)
print("La Especificidad es:",Especificidad)
print("La Precisión es:",Precisión)
print("La Tasa de Error es:",TasaError)
PredNeg=VN/(VN+FP)

6. Determinar las variables con mayor correlación

In [ ]:
import pandas as pd
import numpy as np

# Calcula la matriz de correlación
corr_matrix = XDB.corr()

# Elimina 'games_played' de la matriz de correlación
corr_matrix = corr_matrix.drop('games_played', axis=0).drop('games_played', axis=1)

# Elimina la diagonal (correlación de una variable consigo misma)
corr_matrix = corr_matrix.mask(np.equal(*np.indices(corr_matrix.shape)))

# Encuentra las dos variables con mayor correlación absoluta
max_value = np.nanmax(corr_matrix)  # Valor máximo de correlación
row_max, col_max = corr_matrix[corr_matrix == max_value].stack().idxmax()  # Variables correspondientes

# Imprime el resultado
print(f"Las dos variables con mayor correlación (excluyendo games_played) son: {row_max} y {col_max} con una correlación de {max_value}")

# Para obtener todas las correlaciones por encima de un umbral (por ejemplo, 0.7):
threshold = 0.5  # Define tu umbral
high_corr = corr_matrix[abs(corr_matrix) > threshold].stack().reset_index()
high_corr.columns = ['Variable 1', 'Variable 2', 'Correlación']
print("\nVariables con correlación por encima del umbral:")
print(high_corr)

7. Determinar la categoría a la que pertenece una persona con las siguientes características mediante la utilización del modelo de Clustering y graficar:

In [ ]:
XD1=np.array([(5000000,37,82,245,45,17.8,0.5,1.4,0.4)])

ydn=mkm.predict(XD1)
print("la persona pertenece al cluster:",mkm.predict(XD1))
print("El porcentaje de jugadores por encima del promedio por cluster es es:\n",PJPP)
print("El porcentaje de jugadores por debajo del promedio por cluster es:\n",PJPDP)
XD=np.array(XD)
plt.scatter(XD[:,2],XD[:,3],c=labels,s=50,cmap='viridis')
plt.scatter(XD1[:,2],XD1[:,3],marker='*',s=200,facecolors='red')
plt.scatter(centroids[:,2],centroids[:,3],marker='*',s=200,facecolors='yellow')
plt.scatter(centroids[:,2],centroids[:,3],alpha=0.5,marker='o',s=10000,facecolors='grey')

for i, c in enumerate (centroids):
  plt.text(c[2],c[3],str(i),fontsize=12,color='black',ha='center',va='center')



#**Análisis de Resultados**
Al momento de clasificar los datos, para poder determinar si el número de juegos jugados por jugador de la NBA está por encima del promedio, se utilizaron 3 modelos (Naive Bayes, Árboles de Decisión y K-Means), esto con el fin de encontrar el más eficiente al momento de generar predicciones para este caso de estudio. Además, se tomaron un total de 9 variables de entrada (Salary, Age, Height, Weight, games_started, minutes_per_game, assits, blocks, steals) y una variable explicativa, la cual era (games_played).

1.  Naive Bayes: Al analizar este modelo, se encontró una exactitud del 80%, lo que indica una buena capacidad para clasificar correctamente a los jugadores que se encuentran o no en el promedio. La sensibilidad fue del 82% y la especificidad del 78%, demostrando que el modelo logró identificar correctamente tanto a los jugadores por encima como por debajo del promedio; sin embargo, este asume independencia entre variables, lo cual puede generar que se disminuya la precisión.

	2.  Árboles de decisión: El modelo de árboles de decisión alcanzó una exactitud del 95%, lo que refleja un rendimiento excelente en la clasificación de los jugadores. Además, presentó una tasa de error muy baja del 4.49% aproximadamente. Las métricas de sensibilidad (94.12%) y especificidad (97.14%) son muestran muy buenos resultados, ya que indica que el modelo es muy bueno para reconocer tanto a los jugadores que juegan por encima del promedio y a los que no. También, posee una Precisión muy alta (97.14%), lo cual nos dice que cuando el modelo predice que un jugador juega por encima del promedio, es muy probable que sea correcto.

	3.  K-Means (clustering): Al tratarse de un modelo de clustering, no se evalúa con las mismas métricas que los modelos de clasificación. Se generaron cinco clústeres basados en el porcentaje de jugadores por encima y por debajo del promedio de partidos jugados. Se encontró una exactitud de 73.48%, la cual refleja la limitación que pueda tener al momento de clasificar los datos. Asimismo, las variables con mayor correlación en el modelo son la altura y el peso con un 82.6%.
  
Teniendo en cuenta lo anterior, el modelo de Árboles de decisión es el mejor modelo para implementar en este caso, ya que cuenta con las mejores métricas, haciendo que sean verídicas las predicciones en cuanto a la cantidad de jugadores que están por encima y por debajo del promedio de juegos. El segundo sería el Modelo Naive Bayes, identificando fácilmente que jugadores juegan por debajo del promedio. Por último, el modelo K-means, es el que tuvo un rendimiento regular al momento de hacer las predicciones.
